### PR TITLE
provide unified query resource control interface

### PR DESCRIPTION
**Pull request by jt2594838**
Previous query resource control involves calling methods in several different classes, which increases the difficulty of maintenance.
All methods needed to secure resources are put together into QueryResourceManager and an execution order is provided in its JavaDoc in this PR.

### CODE REVIEW

File: iotdb/src/main/java/org/apache/iotdb/db/query/control/QueryResourceManager.java [hunk 1]
```diff
@@ -34,13 +37,17 @@
...
  * <p>
- * Singleton pattern, to manage all query tokens. Each jdbc query request can query multiple series,
- * in the processing of querying different device id, the <code>FileNodeManager.getInstance().
- * beginQuery</code> and <code>FileNodeManager.getInstance().endQueryForGivenJob</code> must be invoked in the
- * beginning and ending of jdbc request.
+ * QueryResourceManager manages resource (file streams) used by each query job, and assign Ids to the jobs.
+ * During the life cycle of a query, the following methods must be called in strict order:
+ * 1. assignJobId - get an Id for the new job.
+ * 2. beginQueryOfGivenQueryPaths - remind FileNodeManager that some files are being used
+ * 3. (if using filter)beginQueryOfGivenExpression
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/query/control/QueryResourceManager.java]: if using filter, both step 2 and 3 need to be called? or just either of them?

Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/query/control/QueryResourceManager.java]: Both, as you can see, there is no condition on the second step.

File: iotdb/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java [hunk 2]
```diff
@@ -104,6 +104,7 @@
   private ThreadLocal<HashMap<String, QueryDataSet>> queryRet = new ThreadLocal<>();
   private ThreadLocal<ZoneId> zoneIds = new ThreadLocal<>();
   private IoTDBConfig config = IoTDBDescriptor.getInstance().getConfig();
+  private QueryContext context;
```
Comment (little-emotion) [inline iotdb/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java]: Why querycontext is not ThreadLocal like other member variables in this class？

Comment (jt2594838) [reply iotdb/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java]: Fixed.

File: iotdb/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java [hunk 3]
```diff
@@ -618,6 +626,26 @@ public TSFetchResultsResp fetchResults(TSFetchResultsReq req) throws TException
...
+    PhysicalPlan physicalPlan = queryStatus.get().get(statement);
+    processor.getExecutor().setFetchSize(fetchSize);
+
+    QueryContext context = new QueryContext(QueryResourceManager.getInstance().assignJobId());
+    Map<Long, QueryContext> contextMap = contextMapLocal.get();
+    if (contextMap == null) {
+      contextMap = new HashMap<>();
+      contextMapLocal.set(contextMap);
+    }
+    contextMap.put(req.queryId, context);
```
Comment (jixuan1989) [inline iotdb/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java]: so, when client sends request to the server, the client assign an ID. Then, when the server receives a request, it save this ID and then generate an ID in the QueryContext?

**Pull request merged**